WasatchNET/MultiChannelWrapper.cs

```diff
@@ -392,24 +673,62 @@ public async Task<List<ChannelSpectrum>> takeDark(bool sendTrigger=true)
         public void clearDark()
         {
             foreach (var pair in specByPos)
-                pair.Value.dark = null;
+            {
+                var spec = pair.Value;
+                if (spec.multiChannelSelected)
+                    spec.dark = null;
+            }
+        }
+
+        ////////////////////////////////////////////////////////////////////////
+        // Reference
+        ////////////////////////////////////////////////////////////////////////
+
+        /// <summary>
+        /// Takes a spectrum from each spectrometer, and stores it internally as 
+        /// a new reference.
+        /// </summary>
+        /// <returns>The collected references (also stored in Spectrometer)</returns>
+        public async Task<List<ChannelSpectrum>> takeReferenceAsync(bool sendTrigger = true)
+        {
+            clearReference();
+            List<ChannelSpectrum> results = await getSpectraAsync(sendTrigger);
+            foreach (var cs in results)
+                specByPos[cs.pos].reference = cs.intensities;
+            return results;
         }
 
         /// <summary>
-        /// Get a spectrum from one spectrometer.  Assumes HW triggering is disabled, or caller is providing.
+        /// Clears the stored reference from all spectrometers.
         /// </summary>
-        /// <param name="pos">spectrometer position to acquire</param>
-        /// <returns>measured spectrum</returns>
-        public ChannelSpectrum getSpectrum(int pos)
+        public void clearReference()
         {
-            if (!specByPos.ContainsKey(pos))
-                return null;
+            foreach (var pair in specByPos)
+            {
+                var spec = pair.Value;
+                if (spec.multiChannelSelected)
+                    spec.reference = null;
+            }
+        }
 
-            Spectrometer spec = specByPos[pos];
+        ////////////////////////////////////////////////////////////////////////
+        // Post-Processing
+        ////////////////////////////////////////////////////////////////////////
 
-            ChannelSpectrum cs = new ChannelSpectrum(spec);
-            cs.intensities = spec.getSpectrum();
-            return cs;
+        /// <summary>
+        /// It is assumed that either the sample and the reference were both
+        /// dark-corrected, or that neither were.
+        /// </summary>
+        void computeReflectance(Spectrometer spec, ChannelSpectrum cs)

```

Comment:
Interesting...would like to know more if you get the chance. I assume this is sort of a multi channel specific thing?

Reply:
Not really.  ENLIGHTEN has had reflectance/transmission built-in since before I got here...currently implemented in Controller.compute_transmission (the math for reflectance and transmission are identical, and which word you use simply depends on whether the light source is behind the sample (transmission) or beside the spectrometer (reflectance).  It's just (sample / reference), or (sample - dark) / (reference - dark) if you're doing dark correction (as you should :-)